File: android/src/main/kotlin/halo/social/halo_im/xmtp/warppes/ConversationWrapper.kt
package halo.social.halo_im.xmtp.warppes


import android.util.Base64
import com.google.gson.GsonBuilder
import org.xmtp.android.library.Client
import org.xmtp.android.library.Conversation

class ConversationWrapper {

    companion object {
        fun encodeToObj(client: Client, conversation: Conversation): Map<String, Any?> {
            val context = when (conversation.version) {
                Conversation.Version.V2 -> mapOf<String, Any>(
                    "conversationID" to (conversation.conversationId ?: ""),
                    // TODO: expose the context/metadata explicitly in xmtp-android
                    "metadata" to conversation.toTopicData().invitation.context.metadataMap,
                )

                else -> mapOf()
            }
            return mapOf(
                "clientAddress" to client.address,
                "createdAt" to conversation.createdAt.time,
                "context" to context,
                "topic" to conversation.topic,
                "peerAddress" to conversation.peerAddress,
                "version" to "DIRECT",
                "conversationID" to (conversation.conversationId ?: ""),
                "keyMaterial" to (conversation.keyMaterial?.let { Base64.encodeToString(it, Base64.NO_WRAP) } ?: ""),
                "consentProof" to if (conversation.consentProof != null) Base64.encodeToString(conversation.consentProof?.toByteArray(), Base64.NO_WRAP) else null
            )
        }

        fun encode(client: Client, conversation: Conversation): String {
            val gson = GsonBuilder().create()
            val obj = encodeToObj(client, conversation)
            return gson.toJson(obj)
        }
    }
}

File: ios/Classes/Wrappers/ConversationWrapper.swift
//
//  ConversationWrapper.swift
//
//  Created by Pat Nakajima on 4/21/23.
//

import Foundation
import XMTP

// Wrapper around XMTP.Conversation to allow passing these objects back into react native.
struct ConversationWrapper {
    static func encodeToObj(_ conversation: XMTP.Conversation, client: XMTP.Client) throws -> [String: Any] {
        var context = [:] as [String: Any]
        if case let .v2(cv2) = conversation {
            context = [
                "conversationID": cv2.context.conversationID,
                "metadata": cv2.context.metadata,
            ]
        }
        var consentProof: String? = nil
        if (conversation.consentProof != nil) {
            consentProof = try conversation.consentProof?.serializedData().base64EncodedString()
        }
        return [
            "clientAddress": client.address,
            "topic": conversation.topic,
            "createdAt": UInt64(conversation.createdAt.timeIntervalSince1970 * 1000),
            "context": context,
            "peerAddress": try conversation.peerAddress,
            "version": "DIRECT",
            "conversationID": conversation.conversationID ?? "",
            "keyMaterial": conversation.keyMaterial?.base64EncodedString() ?? "",
            "consentProof": consentProof ?? ""
        ]
    }
    
    static func encode(_ conversation: XMTP.Conversation, client: XMTP.Client) throws -> String {
        let obj = try encodeToObj(conversation, client: client)
        let data = try JSONSerialization.data(withJSONObject: obj)
        guard let result = String(data: data, encoding: .utf8) else {
            throw WrapperError.encodeError("could not encode conversation")
        }
        return result
    }
    
    static func encodeToClass(_ conversation: XMTP.Conversation, client: XMTP.Client) async throws -> IMConversation {
        var consentProof: String? = nil
        if (conversation.consentProof != nil) {
            consentProof = try conversation.consentProof?.serializedData().base64EncodedString()
        }
        
        var consentStateStr = ""
        
//        let consentState = try await conversation.consentState();
        
        var groupName = ""
        var groupId = ""
        var groupDescription = ""
        var groupIcon = ""
        var peerAddress = ""
        
        switch conversation {
        case .group(let group):
            groupName = try group.groupName()
            groupId = group.id
            groupDescription = try group.groupDescription()
            groupIcon = try group.groupImageUrlSquare()
            peerAddress = group.id
        default:
            groupName = ""
            groupId = ""
            groupDescription = ""
            groupIcon = ""
            peerAddress = try conversation.peerAddress
        }
        
        let conversation = IMConversation(
            clientAddress: client.address,
            topic: conversation.topic,
            peerAddress: peerAddress,
            version: conversation.versionString(),
            createdAt: Int64(conversation.createdAt.timeIntervalSince1970 * 1000),
            conversationID: conversation.conversationID ?? "",
            keyMaterial: conversation.keyMaterial?.base64EncodedString() ?? "",
            consentProof: consentProof ?? "",
            isAllowed: false,
            isDenied: false,
            groupName: groupName,
            groupDescription: groupDescription, 
            groupId: groupId,
            groupIcon: groupIcon
        );
        return conversation;
    }
}

extension Conversation {
    func versionString() -> String {
        switch self {
        case .v1:
            return "V1"
        case .v2:
            return "V2"
        case .group:
            return "GROUP"
        }
    }
    
}
